File: codes/Garnetwzy/99.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *first = NULL;
    TreeNode *second = NULL;
    TreeNode *pre = NULL;
    void recoverTree(TreeNode* root) {
        help(root);
        if(first && second) {
            int tmp = first->val;
        first->val = second->val;
        second->val = tmp;    
        }
    }
    
    void help(TreeNode* root) {
        if(root == NULL)
            return;
        help(root->left);
        
        if(pre && root->val < pre->val) {
            if(first == NULL) {
                first = pre;
            }
            if(first != NULL) {
                second = root;
            }
        }
        pre = root;
        help(root->right);
    }
};

File: codes/HellDream/99.java
class Solution {
    TreeNode first = null, second = null, pre= null;
    public void recoverTree(TreeNode root) {
        inorder(root);
        if(first!=null && second!=null) swap(first, second);
    }
    public void inorder(TreeNode root){
        if(root==null) return;
        inorder(root.left);
        if(pre==null) pre=root;
        else{
            if(pre.val>root.val){
                if(first==null) first = pre;
                second = root;
            }
            pre = root;
        }
        inorder(root.right);
    }
    
    public void swap(TreeNode first, TreeNode second){
        int tmp = first.val;
        first.val = second.val;
        second.val = tmp;
    }
}